File: BootCamp Java/Date/Date.java
public class Date {
    private int day;
    private int month;
    private int year;
    
    public Date(int day, int month, int year) throws InvalidDateException {
        validate(day, month, year);
        
        this.day = day;
        this.month = month;
        this.year = year;
    }
    public Date() throws InvalidDateException{
        this( 1, 1, 1970);
    }

    private void validate(int day, int month, int year) throws InvalidDateException{
        if ( year < 1 || month < 1 || day < 1 ){
            throw new InvalidDateException("Invalid date");
        }
        if ( year <= 1582 && month <= 10 && day < 15 ){
            throw new InvalidDateException("Try futher date - more than 14.02.1582");
        }
        if ( year < 1582 ){
            throw new InvalidDateException("Try bigger year");
        }
        if ( month > 12 ) {
            throw new InvalidDateException("Try smaller month");
        }
        if ( day > 31 ) {
            throw new InvalidDateException("Try less days");
        }
        if ( month == 2 && day > 29 ) {
            throw new InvalidDateException("Too many days for this month");
        }
        if ( month == 2 && day > 28 && ( year % 4 != 0 || year % 400 != 0 ) ) {
            throw new InvalidDateException("Too many days for this month in this year");
        }
        if ( day > 31 || ( day > 30 && ( month == 4 || month == 6 || month == 9 || month == 11 ) ) )    {
            throw new InvalidDateException("Too many days for this month");
        }
    }


    public int getDay(){
        return this.day;
    }

    public int getMonth() {
        return this.month;
    }

    public int getYear() {
        return this.year;
    }

    public String toString() {
        return day +"."+ month +"."+ year;
    }
}

File: BootCampOOP/Date/Date.cpp
#include "Date.h"

Date::Date(int dayIn, int monthIn, int yearIn){
    validate(dayIn, monthIn, yearIn);
          
    day = dayIn;
    month = monthIn;
    year = yearIn;  
    
}

Date::~Date(){
    
}

void Date::validate(int day, int month, int year){
    bool leapYear = false;
    
    if ( year < 1 || month < 1 || day < 1 ){
        throw InvalidDate ("Дата введена неверно!");
    }
    if ( month > 12 || day > 31) {
        throw InvalidDate ("Дата введена неверно!");
    }
    if ( year % 400 == 0 ){
        leapYear = true;
    }
    else if ( year % 100 == 0) {
        leapYear = false;
    }
    else if ( year % 4 == 0) {
        leapYear = true;
    }
    else {
        leapYear = false;
    }
    
    if (month == 1 ){
        if (day > 31){
            throw InvalidDate ("Дата введена неверно!");
        }
    }
    if (month == 2 && leapYear == false ){
        if (day > 28){
            throw InvalidDate ("Дата введена неверно!");
        }
    }
    else if (month == 2 && leapYear == true ){
        if (day > 29){
            throw InvalidDate ("Дата введена неверно!");
        }
    }
    if (month == 3){
        if (day > 31){
            throw InvalidDate ("Дата введена неверно!");
        }
    }
    if (month == 4){
        if (day > 30){
            throw InvalidDate ("Дата введена неверно!");
        }
    }
    if (month == 5 ){
        if (day > 31){
            throw InvalidDate ("Дата введена неверно!");
        }
    }
    if (month == 6 ){
        if (day > 30){
            throw InvalidDate ("Дата введена неверно!");
        }
    }
    if (month == 7 ){
        if (day > 31){
            throw InvalidDate ("Дата введена неверно!");
        }
    }
    if ( month == 8 ){
        if (day > 31){
            throw InvalidDate ("Дата введена неверно!");
        }
    }
    if ( month == 9 ){
        if (day > 30){
            throw InvalidDate ("Дата введена неверно!");
        }
    }
    if ( month == 10 ){
        if (day > 31){
            throw InvalidDate ("Дата введена неверно!");
        }
    }
    if ( month == 11 ){
        if (day > 30){
            throw InvalidDate ("Дата введена неверно!");
        }
    }
    if ( month == 12 ){
        if (day > 31){
            throw InvalidDate ("Дата введена неверно!");
        }
    }
    
}

int Date::getDay() const{
    return day;
    
}

int Date::getMonth() const{
    return month;
    
}

int Date::getYear() const{
    return year;
    
}

std::ostream& operator<<(std::ostream& out, const Date& date){
    out << date.getDay() << "." << date.getMonth() << "." << date.getYear() << std::endl;
}